common/dbconfig.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig(new DBConfig()) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig, newConfig, std::memory_order_release);
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+std::shared_ptr<const DBConfig> DBConfigManager::parseConfig(const Json::Value& root) {
+  static const std::string ACK_MODE = "ack_mode";
+  static const std::string DATASET = "dataset";
+
+  DBConfig newDBConfig;
+
+  if (root.isMember(DATASET)) {
+    // parse dataset specific config
+    auto jsonDataSetConfigs = root[DATASET];
+
+    for (const auto& dataset : jsonDataSetConfigs.getMemberNames()) {
+      Config config;
+      const auto& jsonDataSetConfig = jsonDataSetConfigs[dataset];
+      bool isSet = false; /* creating this for future multiple assigns */
+
+      if(jsonDataSetConfig.isMember(ACK_MODE)) {
+        config.replication_mode = jsonDataSetConfig[ACK_MODE].asUInt();
+        isSet = true;
+      }
+
+      if (isSet) {
+        newDBConfig.dataSetConfigMap[dataset] = config;
+      }
+    }
+  }
+  return std::make_shared<const DBConfig>(std::move(newDBConfig));
+}
+
+/* conf getter functions */
+const Config* DBConfigManager::getConfig(const std::string& dbName) const {
+  auto dsName = DbNameToSegment(dbName);
+  const auto& dsConfigMap = std::atomic_load_explicit(&dbConfig, std::memory_order_acquire)->dataSetConfigMap;
+  const auto dsIter = dsConfigMap.find(dsName);
+  if (dsIter == dsConfigMap.end()) {
+    return nullptr;
+  } else {
+    return &(dsIter->second);
+  }
+}
+
+uint DBConfigManager::getReplicationMode(const std::string& dbName, uint defValue) const {
+  const auto* config = getConfig(dbName);
+  if (config != nullptr) {
+    return config->replication_mode;
```

Comment:
is there still a race condition here, since `getConfig` returns a pointer here there is no locking in reading its `replication_mode` --- it could have been mutated after `getConfig` is retuned

Reply:
True, We can leave that as opposed to using a lock/unlock by all threads ..

Reply:
hmm, we just need read lock here, so no contention for concurrent reads between threads. Only lock contention will happen is when there is a concurrent write (i.e. when config gets updated), which should rarely happen. 

(correctness over performance)

Reply:
Fixed this by making a copy of the shared pointer in getConfig and returning a shared_ptr of Config..